### ISSUE TITLE
iris.analysis.interpolate failure

### ISSUE THREAD

**Issue by djkirkham** [SciTools/iris#2116]
A user is reporting that some library code based on Iris is failing in  v1.10 now that iris.analysis.interpolate has been deprecated. The code in question has the following form:

```
import iris.analysis
...
iris.analysis.interpolate.linear(...)
```

I know this is incorrect usage (thanks @dkillick), but given that it works in v1.9.2, is there a case for making sure it works in v1.10?

**Comment by dkillick** [comment]
There is, it's...

```python
from iris.analysis.interpolate import linear

...
linear(...)
```

:wink: 

Sometimes moving forwards means breaking stuff. Typically this should be done in a major version, but I think we can get away with it in this case, given that this was never intentional behaviour on Iris' part, and should not have been treated by end users as such.

**Comment by djkirkham** [comment]
This puts the user in question in a bit of a bind. He can't control which version of Iris is used, so his library will suddenly break for all his users until he can get a fix out.

I would normally agree with you that we shouldn't worry about this sort of thing, and if the code in question were, say, using a private variable I wouldn't be in favour of keeping it working. But this case is kind of a Python gotcha. I didn't even know it was incorrect usage myself.

**Comment by ajdawson** [comment]
>This puts the user in question in a bit of a bind. He can't control which version of Iris is used, so his library will suddenly break for all his users until he can get a fix out.

This is an unfortunate scenario for the user in question, but this does happen. The fix for the library in question seems very simple, and this is exactly what a bugfix release is for.

Since we've deprecated this module we shouldn't be importing it into the `iris.analysis` namespace for backwards compatibility, after all this will break end user code in the next release when the interpolate module disappears.

**Comment by ajdawson** [comment]
There are also bigger problems on the horizon for this user, of course... The library in question will need to be updated for compatibility with iris 2.0 when there is no `iris.analysis.interpolate` module at all. Now might be the time to guide them towards using the `interpolate` and/or `regrid` methods on the cube to ensure they don't have similar breakage when their end users upgrade to iris 2.0.

**Comment by dkillick** [comment]
> Now might be the time to guide them towards using the `interpolate` and/or `regrid` methods on the cube

:+1:

**Comment by djkirkham** [comment]
I can see I'm fighting a losing battle here, but my suggestion is only to keep the user's code working in the same way `import iris.analysis.interpolate` will still work, and with the same deprecation warning. Both will break in the next release, but at least the user would have some warning.

**Comment by ajdawson** [comment]
@djkirkham - please submit a PR with your proposed technical implementation to the 1.10 release branch and it will be reviewed. Since 1.10 is in release candidate already this will likely need to be done soon.

**Comment by marqh** [comment]
> A user is reporting that some library code based on Iris is failing in v1.10 now that iris.analysis.interpolate has been deprecated. The code in question has the following form:
> 
> import iris.analysis
> ...
> iris.analysis.interpolate.linear(...)
> 
> I know this is incorrect usage (thanks @dkillick), but given that it works in v1.9.2, is there a case for making sure it works in v1.10?

I agree with Dan about this.  I think that we should be preserving import paths wherever possible for minor releases.
@djkirkham I'm with @ajdawson here; please put up a PR, targeting the 1.10.x branch which restores the ability to get to things such as iris.analysis.interpolate.linear by importing iris.analysis

Please liaise with @pp-mo on how this may be achieved without raising deprecation warnings, I suspect this is why the import path was was taken out as part of the deprecation.  
Perhaps there is a neat way to work within these constraints.

@dkillick 
> Sometimes moving forwards means breaking stuff. Typically this should be done in a major version, but I think we can get away with it in this case, given that this was never intentional behaviour on Iris' part, and should not have been treated by end users as such.

I don't agree; Iris has changed and it is hard for a user to understand the import path changes that are made.
I don't think intent is the issue, it is the result and whether it is easy to maintain the behaviour.

**Comment by marqh** [comment]
closed by #2121
